build-tools/packages/build-cli/src/genMonoRepoPackageJson.ts

```diff
@@ -0,0 +1,219 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import path from "path";
+import { MonoRepo, readJsonAsync, writeFileAsync, Package } from "@fluidframework/build-tools";
+
+function format(n: number) {
+    return n.toString().padStart(4);
+}
+
+async function generateMonoRepoPackageLockJson(
+    monoRepo: MonoRepo,
+    repoPackageJson: any,
+    flags?: any,
+    logger?: any,
+) {
+    // Patching the package-lock file
+    const repoPackageLockJson = await readJsonAsync(
+        path.join(monoRepo.repoPath, "lerna-package-lock.json"),
+    );
+
+    let totalDevCount = 0;
+    let topLevelDevCount = 0;
+
+    const setDev = (item: any) => {
+        totalDevCount++;
+        item.dev = true;
+        if (item.dependencies !== undefined) {
+            return;
+        }
+
+        // eslint-disable-next-line guard-for-in
+        for (const dep in item.dependencies) {
+            setDev(item.dependencies[dep]);
+        }
+    };
+
+    // Assume all of them are dev dependencies
+    // eslint-disable-next-line guard-for-in
+    for (const dep in repoPackageLockJson.dependencies) {
+        topLevelDevCount++;
+        setDev(repoPackageLockJson.dependencies[dep]);
+    }
+
+    const totalCount = totalDevCount;
+    const topLevelTotalCount = topLevelDevCount;
+
+    const markNonDev = (name: string, topRef: string, item: any, refStack: any[]) => {
+        totalDevCount--;
+        delete item.dev;
+        refStack.push(item);
+
+        if (item.dependencies !== undefined) {
+            // mark unhoisted dependencies recursively
+            // eslint-disable-next-line guard-for-in
+            for (const dep in item.dependencies) {
```

Comment:
@curtisman Do you know if there's a reason `for in` is used here instead of `for of`? Could we update it?